### ISSUE TITLE
Response has invalid UTF-8

### ISSUE THREAD

**Issue by manishrjain** [dgraph-io/dgraph4j#29]
Posted via: https://github.com/dgraph-io/dgraph/issues/1789

@rajeshkmindix

Hi,

I get the following error when trying to insert some nodes into dgraph.

com.google.protobuf.InvalidProtocolBufferException: Protocol message had invalid UTF-8.

I narrowed it down to

```java
ManagedChannel channel = ManagedChannelBuilder.forAddress("localhost", 9080).usePlaintext(true).build();
DgraphGrpc.DgraphBlockingStub blockingStub = DgraphGrpc.newBlockingStub(channel);
DgraphClient dgraphClient = new DgraphClient(Collections.singletonList(blockingStub));

for(int i = 0; i < 1000; i++) {
  JsonObject json = new JsonObject();
  json.addProperty("id", i);
  json.addProperty("name", "abcdefgh" + i);

  System.out.println("for " + i);
  Mutation mu =
          Mutation.newBuilder()
                  .setCommitImmediately(true)
                  .setSetJson(ByteString.copyFromUtf8(json.toString()))
                  .build();
  dgraphClient.newTransaction().mutate(mu);
}
```

After inserting around 100 to 150 nodes. I get the above error. If I rerun the above code, I get the same error for the first record itself. Every subsequent mutation request gets the same error.

I installed using
curl https://get.dgraph.io -sSf | bash

Ran it using
dgraph zero --port_offset -2000 dgraph server --memory_mb 2048 --zero localhost:5080
Need help with this.

Thanks in advance,
Rajesh

**Comment by deepakjois** [comment]
This report looks similar to #25. Investigating now.

**Comment by deepakjois** [comment]
So this looks like a bug in the server, more specifically the protocol buffers definition for `TxnContext`.

Right now it is like this:

```
message TxnContext {
  uint64 start_ts = 1;
  uint64 commit_ts = 2;
  bool aborted = 3;
  repeated string keys = 4;
  LinRead lin_read = 13;
}
```
note the `repeated string keys = 4`. This means that (according to the proto language spec) the keys field must be a valid UTF-8 string . But the server currently sends a stream of bytes without bothering to check if it is valid UTF-8. The java client (autogenerated code) tries to decode the stream as valid UTF-8 and throws an error if it is not able to decode it as valid UTF-8.

The fix is to base64 encode the strings before sending them, so that they are guaranteed to be valid UTF-8.

This will require a fix to the server. I will keep this issue updated. (cc: @pawanrawal)

**Comment by deepakjois** [comment]
This is now fixed in https://github.com/dgraph-io/dgraph/commit/f27fed9e2547935b8d60a9211a543f455307461f

We will be doing a new `v0.9.2` server release today which will contain this fix. You can try to run your code against that. I believe it will work.

Meanwhile I have added a test for this case here: 4562c788f7868070f239883a5191997980fcec42

**Comment by deepakjois** [comment]
Dgraph v0.9.2 is released, which should work with the client at v0.9.2.

Closing now, feel free to reopen if there is any other issues.
